storage: serve reads based on closed timestamps

tbg opened the pull request:
This "finishes" hooking up the storage side of closed timestamps. After
checking their lease and deciding that the lease does not allow serving
a read, replicas now check whether they can serve the batch as a
follower read. This requires that the range is epoch based, that
appropriate information is stored in the closed timestamp subsystem, and
finally that the cluster setting to enable this is set.

Added a test that verifies that a test server will serve follower reads
(directly from the replicas, without routing through DistSender).

Introducing machinery at the distributed sender to actually consider
routing reads to follower replicas is the next step.

TODO: take perf numbers before/after this change to verify that there
isn't a noticeable regression.

Touches #16593.

Release note: None

tschottdorf commented:
Heh, interesting test flake here:

```
[15:57:34] :		 [Maybe stressrace pull request] --- FAIL: TestClosedTimestampCanServe (77.43s)
[15:57:34] :		 [Maybe stressrace pull request] 	closed_timestamp_test.go:181: replica corruption (processed=true): could not read from AbortSpan: cannot read undeclared span /Local/RangeID/24/r/AbortSpan/"adb6f9f3-4385-45b6-b55c-4e5efb907ef6"
[15:57:34] :		 [Maybe stressrace pull request] 		declared:
[15:57:34] :		 [Maybe stressrace pull request] 		0 1: /Local/RangeID/0/r/AbortSpan/"adb6f9f3-4385-45b6-b55c-4e5efb907ef6"{-}
[15:57:34] :		 [Maybe stressrace pull request] 		1 0: /Table/53{-/NULL}
```

I bet that has nothing to do with my change, but I'll fix it anyway.

nvb reviewed (changes requested):
> Reviewed 13 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28091)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/closed_timestamp_test.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq78jG3FDLlyGeVciK:-LIq78jG3FDLlyGeVciL:b-rznlgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closed_timestamp_test.go#L144)):*
> ```Go
> 	baRead.Timestamp = hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
> 
> 	testutils.SucceedsSoon(t, func() error {
> ```

Add a comment to this. It's the exciting part!
___
*[pkg/storage/replica.go, line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq37ZB5EQfj2orK5Qk:-LIq37ZB5EQfj2orK5Ql:buqk89k) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L138)):*
> ```Go
> // setting.
> var FollowerReadsEnabled = settings.RegisterBoolSetting(
> 	"server.closed_timestamp.follower_reads_enabled",
> ```

nit: `kv.closed_timestamp.follower_reads_enabled` to mirror the other `kv` settings?
___
*[pkg/storage/replica.go, line 2851 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq3ZB18Jzt5qwBM6-n:-LIq3ZB18Jzt5qwBM6-o:b-qjf72f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L2851)):*
> ```Go
> 	if ba.ReadConsistency.RequiresReadLease() {
> 		if status, pErr = r.redirectOnOrAcquireLease(ctx); pErr != nil {
> 			if lErr, ok := pErr.GetDetail().(*roachpb.NotLeaseHolderError); ok &&
> ```

Previous versions of this code avoided the lease acquisition attempt entirely when a follower read could be served. What's the justification for this approach? Do we do this just to get the leaseholder's epoch?
___
*[pkg/storage/replica.go, line 2877 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq4AA77ZWwoDuQqQMe:-LIq4AA77ZWwoDuQqQMf:b-5els5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L2877)):*
> ```Go
> 				// TODO(tschottdorf): once a read has been proven servable, we can serve all reads
> 				// for the same or lower timestamps as long as the lease doesn't change hands (or
> 				// epoch).
> ```

Even if the lease changes hands or epoch, why would we need to stop serving follower reads for the same timestamp? Is this an artifact of the implementation or a fundamental limitation?
___
*[pkg/storage/replica.go, line 3415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq4uUO6Ozjj5-yKiO9:-LIq4uUO6Ozjj5-yKiOA:b-lxifwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L3415)):*
> ```Go
> 	// that enough? If the request isn't tracked, it wouldn't have an Untrack
> 	// method. We only explicitly pass nil in some low level tests that don't
> 	// care about closed timestamps.
> ```

We could eliminate all of the cases where `proposal.Untrack` is nil and make it invalid.
___
*[pkg/storage/replica.go, line 3417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq586t7kwQAY81nhze:-LIq586t7kwQAY81nhzf:b-x10rkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L3417)):*
> ```Go
> 	// care about closed timestamps.
> 	if proposal.Untrack != nil {
> 		proposal.Untrack(proposal.ctx, r.RangeID, ctpb.LAI(proposal.command.MaxLeaseIndex))
> ```

Since we're pulling this all off the `ProposalData` object, I wonder if it's possible to call this all the way back up in `tryExecuteWriteBatch` right after we call `propose`.
___
*[pkg/storage/store.go, line 3793 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq5gAV6kHBy2pockqw:-LIq5gAV6kHBy2pockqx:b-2bipci) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/store.go#L3793)):*
> ```Go
> 	for nodeID, live := range livenessMap {
> 		if live {
> 			s.cfg.ClosedTimestamp.Clients.EnsureClient(nodeID)
> ```

This is quick if the client connection already exists, right?
___
*[pkg/storage/closedts/container/noop.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq61hr74JaLxnk6hrp:-LIq61hr74JaLxnk6hrq:basd7c9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closedts/container/noop.go#L48)):*
> ```Go
> 			Dialer:  noopEverything{},
> 		},
> 		Tracker:  noopEverything{},
> ```

Do we need all this if we're going to add a `noop` flag to `Container` anyway?
___
*[pkg/storage/closedts/provider/provider_test.go, line 284 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq6UzlAf5VLFOW-I2x:-LIq6UzlAf5VLFOW-I2y:b-5joz1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closedts/provider/provider_test.go#L284)):*
> ```Go
> 			ch := make(chan ctpb.Entry, 3)
> 			p.Subscribe(ctx, ch)
> 			for range ch {
> ```

Add a comment like "Read from channel until Provider is Stopped and the channel is closed."
___
*[pkg/storage/closedts/storage/storage.go, line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq7eKl5jqPGzSeafJF:-LIq7eKl5jqPGzSeafJG:b-ui5lx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closedts/storage/storage.go#L141)):*
> ```Go
> 	}
> 
> 	shouldPrint := map[roachpb.NodeID]struct{}{}
> ```

nit: give the map an initial cap of `len(nodes)`. Also maybe avoid the allocation entirely if `len(nodes) == 0`
___
*[pkg/storage/closedts/storage/storage.go, line 156 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq7ZCI77xQ1M9ODkV_:-LIq7ZCI77xQ1M9ODkVa:b-w37tj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closedts/storage/storage.go#L156)):*
> ```Go
> 	var buf bytes.Buffer
> 	for i := range sl {
> 		if len(shouldPrint) > 0 {
> ```

Can we move this up to where we populate `sl` in `ms.m.Range`?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28091)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/closed_timestamp_test.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq78jG3FDLlyGeVciK:-LIqq9l131--zgWpT29t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closed_timestamp_test.go#L144)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a comment to this. It's the exciting part!
</blockquote></details>

Done.
___
*[pkg/storage/replica.go, line 2851 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq3ZB18Jzt5qwBM6-n:-LIqqRH4AteSb479_oAP:b-bki14z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L2851)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Previous versions of this code avoided the lease acquisition attempt entirely when a follower read could be served. What's the justification for this approach? Do we do this just to get the leaseholder's epoch?
</blockquote></details>

There's no particular reason except the simplicity (and the desire to avoid pessimizing the regular read path). We do need the leaseholder and epoch, so I'm not sure how much we save by introducing extra code. Either way, that's not something I'm planning to look at before the 2.2 cycle.
___
*[pkg/storage/replica.go, line 2877 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq4AA77ZWwoDuQqQMe:-LIqql2z6sXGSBMbaWMF:bpr9uct) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L2877)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Even if the lease changes hands or epoch, why would we need to stop serving follower reads for the same timestamp? Is this an artifact of the implementation or a fundamental limitation?
</blockquote></details>

Neither! The comment was unnecessarily pessimistic. Once you've served at T, you can serve at T no matter what. But I'm not planning on introducing any bells and whistles here before the 2.2 feature dev cycle.
___
*[pkg/storage/replica.go, line 3415 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq4uUO6Ozjj5-yKiO9:-LIqqwDAEp78IyC-KtCO:b3m9ip8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L3415)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We could eliminate all of the cases where `proposal.Untrack` is nil and make it invalid.
</blockquote></details>

This is gone now thanks to your suggestion below.
___
*[pkg/storage/replica.go, line 3417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq586t7kwQAY81nhze:-LIqqykg-XvvYmluXUdg:b-l4um6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/replica.go#L3417)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Since we're pulling this all off the `ProposalData` object, I wonder if it's possible to call this all the way back up in `tryExecuteWriteBatch` right after we call `propose`.
</blockquote></details>

Done, good idea!
___
*[pkg/storage/store.go, line 3793 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq5gAV6kHBy2pockqw:-LIqr-YgCYAurH-S2G8e:br81w24) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/store.go#L3793)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is quick if the client connection already exists, right?
</blockquote></details>

Yeah, this call is cheap, even if it does not exist.
___
*[pkg/storage/closedts/container/noop.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq61hr74JaLxnk6hrp:-LIqr20365dXwBWAq4sv:bdho4tv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closedts/container/noop.go#L48)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need all this if we're going to add a `noop` flag to `Container` anyway?
</blockquote></details>

The noop flag had to be added to avoid starting goroutines that would never be stopped. Components of the container may still receive calls, and I don't want to have to check this noop flag all over the place. This could all be cleaned up a little bit, but not worth the time right now.
___
*[pkg/storage/closedts/provider/provider_test.go, line 284 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq6UzlAf5VLFOW-I2x:-LIqr9UB28XhWaWuL8eZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closedts/provider/provider_test.go#L284)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add a comment like "Read from channel until Provider is Stopped and the channel is closed."
</blockquote></details>

Done.
___
*[pkg/storage/closedts/storage/storage.go, line 156 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIq7ZCI77xQ1M9ODkV_:-LIqrAMkEU0A_h0kPKZL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c023dd2554345d4970d18e08bbbfab7821cfa232/pkg/storage/closedts/storage/storage.go#L156)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we move this up to where we populate `sl` in `ms.m.Range`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

tschottdorf commented:
After fixing a pretty bad bug (node connecting to itself, leading to larger and larger ct updates):

```
==== closedts ===
=== RUN   kv0/encrypt=false/nodes=3
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        8079105        13465.1     14.3     10.5     41.9     60.8    121.6  write

==== master ===
=== RUN   kv0/encrypt=false/nodes=3
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        8120248        13533.7     14.2     10.5     41.9     60.8    117.4  write
```

Running kv95 now, and I'll also re-run kv0 a few times to see if there's a tendency one way or another.

tschottdorf commented:
On kv95 too, closedts is a little slower. Have to rerun these a few times to confirm.

```
==== closedts ===
=== RUN   kv95/encrypt=false/nodes=3
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
I180802 11:23:01.559550 1 workload/workload.go:382  starting 1000 splits
I180802 11:23:05.798417 1 workload/workload.go:439  finished 1000 of 1000 splits
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       18894632        31490.9      5.2      4.2     16.3     31.5    159.4  read
  600.0s        0         992957         1654.9     17.7     15.2     41.9     58.7    218.1  write
==== master ===
=== RUN   kv95/encrypt=false/nodes=3
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
I180802 11:33:22.445275 1 workload/workload.go:382  starting 1000 splits
I180802 11:33:24.109238 1 workload/workload.go:439  finished 1000 of 1000 splits

--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       18945511        31575.7      5.1      4.2     16.3     31.5    176.2  read
  600.0s        0         998841         1664.7     17.7     15.2     41.9     58.7    234.9  write
```

tschottdorf commented:
One more, similar result. We're losing a little bit. Nothing stood out on the profiles, but it's not dramatic to begin with.

```
=== RUN   kv0/encrypt=false/nodes=3 (ct)
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        8547094        14245.1     13.5     10.0     41.9     58.7    142.6  write

=== RUN   kv0/encrypt=false/nodes=3 (master)
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        8684965        14474.9     13.3     10.5     39.8     58.7    121.6  write

=== RUN   kv95/encrypt=false/nodes=3 (ct)
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       20422042        34036.5      4.7      3.7     15.7     32.5    109.1  read
  600.0s        0        1074591         1791.0     17.3     14.7     41.9     58.7    184.5  write
=== RUN   kv95/encrypt=false/nodes=3 (master)
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       20583437        34305.6      4.7      3.7     15.7     31.5    117.4  read
  600.0s        0        1082450         1804.1     17.0     14.2     41.9     56.6    159.4  write
```

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-:-LIuq2xs-qNMQ0Y-X-Rd:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica.go, line 3560 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIupgYq6O5noEpEMR7D:-LIupgYq6O5noEpEMR7E:b-onpxzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/405c4ff3b296c3990c3baf46b3474269162c4926/pkg/storage/replica.go#L3560)):*
> ```Go
> 	// closed timestamp tracker is acquired. This is better anyway; right now many
> 	// commands can evaluate but then be blocked on quota, which has worse memory
> 	// behavior.
> ```

That's tricky though because we key into the quota pool using the proposal size itself.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
These performance deltas are unfortunate, but not overly so. Do either of the cluster settings introduced here make a difference? For instance, would dropping the `kv.closed_timestamp.close_fraction` to `0.02` help us lose less ground?

tbg reviewed (commented):
I'll play around with it a bit more. It'll be a good exercise to track down the source of the difference, though it's pretty small and it may prove difficult.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-:-LIv5q6c-2djw8pwjiCe:b2tiuc8)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/storage/replica.go, line 3560 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28091#-LIupgYq6O5noEpEMR7D:-LIv5yuE-tn5sN8YUnCa:b-xf1rjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/405c4ff3b296c3990c3baf46b3474269162c4926/pkg/storage/replica.go#L3560)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

That's tricky though because we key into the quota pool using the proposal size itself.
</blockquote></details>

That's true, though something could still be done there. I'm much more worried about blocking the closed timestamp though, so kicking a proposal back up above the tracker reference when it finds out that it's blocked on the pool is more important. Naively that would discard the result of the evaluation, which would be unfortunate. Perhaps we can release the proposal when we block on the pool and check that we can get a new reference with a compatible timestamp after (and only if not kick it back up). Sounds kinda complex too, though.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: serve reads based on closed timestamps

This "finishes" hooking up the storage side of closed timestamps. After
checking their lease and deciding that the lease does not allow serving
a read, replicas now check whether they can serve the batch as a
follower read. This requires that the range is epoch based, that
appropriate information is stored in the closed timestamp subsystem, and
finally that the cluster setting to enable this is set.

Added a test that verifies that a test server will serve follower reads
(directly from the replicas, without routing through DistSender).

Introducing machinery at the distributed sender to actually consider
routing reads to follower replicas is the next step.

TODO: take perf numbers before/after this change to verify that there
isn't a noticeable regression.

Touches #16593.

Release note: None

tschottdorf commented:
Ran another time (after fixing some tiny perf things), this time it looks like my branch came out ahead twice (with the same `master`). That's probably due to noise, but shows that there's no real perf regression.

```
=== RUN   kv0/encrypt=false/nodes=3 (closedts)
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        7733087        12888.4     14.9     11.0     44.0     65.0    159.4  write

=== RUN   kv0/encrypt=false/nodes=3 (master)
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        7696339        12827.2     15.0     11.0     44.0     65.0    285.2  write

=== RUN   kv95/encrypt=false/nodes=3 (closedts)
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       16630597        27717.3      6.0      5.0     19.9     35.7    159.4  read
  600.0s        0         876423         1460.7     16.7     14.2     41.9     56.6    260.0  write
=== RUN   kv95/encrypt=false/nodes=3 (master)
> /home/tschottdorf/go/bin/roachprod list tobias-ct --json
> /home/tschottdorf/go/bin/roachprod stop tobias-ct:1-4
--
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0       16506003        27509.9      6.1      4.7     19.9     35.7    192.9  read
  600.0s        0         868926         1448.2     16.8     14.2     41.9     56.6    218.1  write
```

Also fixed the test failure (which was deterministic under race and caused by an error in the test).

tschottdorf commented:
@nvanbenschoten merging this as you suggested privately, but please take a look at the new diffs to make sure I didn't introduce any easter eggs there.

bors r=nvanbenschoten
